File: two-pointers/rotate-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def rotateRight(self, head, k):
        """
        :type head: Optional[ListNode]
        :type k: int
        :rtype: Optional[ListNode]
        """
        if(head == None or head.next == None ):
            return head
        length = 1
        tail = head
        while(tail.next != None):
            tail = tail.next
            length += 1
        k = k % length
        if(k==0):
            return head
        tail.next = head
        n = length - k
        new_tail = head
        for i in range(n-1):
            new_tail = new_tail.next
        head = new_tail.next
        new_tail.next = None
        return head

File: two-pointers/rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if(k==0 || head == nullptr || head->next == nullptr){
            return head;
        }
        int n = 0;
        ListNode* temp = head;
        while(temp != nullptr){
            temp = temp->next;
            n++;
        }
        if(k == n){
            return head;
        }
        k = k%n;
        if(k == 0){
            return head;
        }
        ListNode* after = head;
        ListNode* before = nullptr;
        ListNode* temp1 = after;
        for(int i = 0; i < n-k-1; i++){
            temp1 = temp1->next;
        }
        before = temp1->next;
        temp1->next = nullptr;
        temp = before;
        while(temp->next != nullptr){
            temp = temp->next;
        }
        temp->next = after;
        return before;
    }
};